### ISSUE TITLE
Wrong suggestion R1706: Consider using ternary 

### ISSUE THREAD

**Issue by MatthijsBlom** [PyCQA/pylint#2058]
Consider the line
```python
original = P and Q or R
```
For this line, Pylint will suggest `R1706 Consider using ternary (Q if P else R)`. This is a mistake, since
```python
P, Q, R = True, False, True
original   = P and Q or R   # True
suggestion = Q if P else R  # False
```
As such, the suggested expression is not equivalent to the original.

The above particular assignment of P, Q, R is the only problematic one: for all other (boolean) assignments of P, Q, and R, the original and suggestion are the same.

### pylint --version output

No config file found, using default configuration
pylint 1.8.4,
astroid 1.6.1
Python 3.6.5 (v3.6.5:f59c0932b4, Mar 28 2018, 16:07:46) [MSC v.1900 32 bit (Intel)]

**Comment by PCManticore** [comment]
Thanks for creating this issue! I agree that the suggestion is wrong. Curious though why the linked fix didn't actually solve this problem.

**Comment by PCManticore** [comment]
@pheanex Yeah, I can see where the bug is. `truth_value` in that example is a `Name` is one of the nodes that don't have a `bool_value()` implementation by default, using the default one which returns `Uninferable`. the fix here should be simple, first infer the `truth_value` object and then call `bool_value()`on it, instead on calling it on the `truth_value`.

Feel free to submit a PR if you have some time!

**Comment by pheanex** [comment]
I may be mistaken, but after thinking on this: Don't we also have to change the suggestion for
"consider-using-ternary" from `Q if P else R` to `Q if P and Q else R` in general, because of this special case?
My thinking is as follows:
* If we can infer the value of "condition"/`P` to either True or False, then we can emit a "simplify-boolean-expression".
* if we can not infer the value safely, we have to emit `Q if P and Q else R`, because for `P, Q, R = True, False, True` the expression `P and Q or R` is still only equal to `Q if P and Q else R` and not `Q if P else R`

**Comment by PCManticore** [comment]
I agree with the first point, that we can just suggest `simplify-boolean-expression` instead, but I'm not sold on the other one. My reasoning is that it is embedding an edge case in a general ternary operation suggestion. Admittedly it would be more correct, but it will become annoying to double check the `Q` condition when you only intended to check that `P` has a truth value or not.

**Comment by pheanex** [comment]
I would love to suggest `Q if P else R` as well, but how can we do this without suggesting something wrong as well? => How do you want to check if we can safely emit "consider-using-ternary" if we can not infer `P`?

**Comment by PCManticore** [comment]
Simple, even if it is an `Uninferable`, we emit `Q if P else R`. If the original `P and Q or R`, where our proposed suggestion was wrong, would have tripped a human reviewer as well, not just a linter, so I don't worry about having false positives here. Also imagine that most of the time we're not able to infer the values involved in an expression, either they come from arguments and we didn't see the call site yet, either they are imported from a module we cannot access etc.

**Comment by MatthijsBlom** [comment]
[datetime source code]: https://github.com/python/cpython/blob/3.7/Lib/datetime.py#L38 "cpython/datetime.py at 3.7 · python/cpython"

This is the exact line in which I encountered this bug:

```python
is_leap_year = year % 4 == 0 and year % 100 != 0 or year % 400 == 0
```

Incidentally, an almost identical expression can be found in the [datetime source code][]. There, however, parentheses are used to group on `or` instead of `and`. (I am not sure why?) This does not change the semantics (in this specific instance), but it does withhold Pylint from emitting R1706.

( I thought I remembered that the parentheses do not matter, but it appears I was wrong, as both 2.0.0.dev2 (on 3.7.0) and 1.9.2 (on 3.6.6) behave as described above. Did something change? )

---

In addition to @pheanex's suggestion I would like to point out a similar alternative:
```python
original = P and Q or R
pylint   = Q if P else R         # Wrong!
pheanex  = Q if P and Q else R
similar  = Q or R if P else R    # <-- This one
```
I think both `pheanex` and `similar` are somewhat nice, but doubt they are so nice that we should urge others to prefer them over the original. To me, both are less enlightening than the original.

---

@PCManticore I am not sure I understand your comment correctly. Are you suggesting this bug need not be fixed?

With regard to confusion: should a linter not, at te very least, not _add_ to the confusion -- not even under already confusing circumstances?

A possible drawback to not fixing this bug is that even if an original author has rightfully decided to not heed Pylint's warning, a later author might decide otherwise and introduce a bug. I am not sure how objectionable this is, however, as I do not code for a living.

---

Thinking about alternatives for the wrong suggestion has made me wonder in what kind of situation one might wish to be confronted with a `consider-using-ternary`, if ever. I don't think the (above) one concrete use case I have encountered so far is one of them.

Rather than ask _How do we salvage this feature?_ I would like to ask **When do we actually _want_ this feature?** Perhaps it is best to simply ditch it?
